src/atc/eh/EventHubCapture.py

```diff
@@ -0,0 +1,210 @@
+
+        # return the partition in path form so that the repair can continue
+        return partition_path
+
+    def _get_max_partition(self):
+        """get the current highest partition either
+        1. from memory
+        2. read the partition table
+        3. initialize from disk
+        """
+        if self.max_partition is None:
+            # memory access failed.
+            # try reading partition table
+            try:
+                max_p = (
+                    Spark.get()
+                    .sql(f"SHOW PARTITIONS {self.name}")
+                    .groupBy()
+                    .agg(f.max("partition"))
+                    .collect()
+                )[0][0]
+                # value will be None if there is no value
+            except AnalysisException:
+                # no such table. Better create one
+                self._create_table()
+                # new table has no partitions
+                max_p = None
+
+            if not max_p:
+                # reading partition metadata failed.
+                # initialize first partition from disk.
+                max_p = self._discover_first_partition()
+
+            # we got the partition in the form y=2022/m=04/d=04
+            # extract the parts
+            pat = re.compile("/".join(f"{c}=(?P<{c}>\\d+)" for c in self.partitioning))
+            match = pat.match(max_p)
+            self.max_partition = datetime(
+                year=int(match.group("y")),
+                month=int(match.group("m")),
+                day=int(match.group("d")),
+                hour=int(match.group("h")) if "h" in self.partitioning else 0,
+            )
+        return self.max_partition
+
+    def _repair_partitioning(self):
+        self._get_max_partition()
+
+        # compare with current wall-clock to see if partitions are missing
+        now = datetime.utcnow().replace(minute=0, second=0, microsecond=0)
+        if "h" not in self.partitioning:
+            now = now.replace(hour=0)
+
+        additions = []
+        while self.max_partition < now:
+            # we know that the partition table is behind
+            next_partition = self.max_partition + self.partition_delta
+            partition_spec = (
+                f"y='{next_partition.year:04}', "
+                f"m='{next_partition.month:02}', "
+                f"d='{next_partition.day:02}'"
+            )
+            if "h" in self.partitioning:
+                partition_spec += f", h='{next_partition.hour:02}'"
+            additions.append(f"PARTITION ({partition_spec})")
+
+        while additions:
+            # We need to add partitions
+            parts_per_batch = min(len(additions), 24)
+            print(f"adding {parts_per_batch} partitions")
+
+            # add partitions in batch to prevent an sql line with hundreds of parts
+            batch = [additions.pop() for _ in range(parts_per_batch)]
+            Spark.get().sql(f"ALTER TABLE {self.name} ADD " + " ".join(batch))
```

Comment:
Why is the parts_per_batch 24?

Reply:
for hourly partitions, it adds a day at a time. For if that number is good enough for hourly, why not keep it for the other case.